### PR TITLE
RavenDB-16614 Test for "Atomic read write in cluster wide transaction without conflicts"

### PR DESCRIPTION
**Pull request by haludi**
https://issues.hibernatingrhinos.com/issue/RavenDB-16614

### CODE REVIEW

**Comment by ayende** [comment]
Isn't this supposed to be on top of: https://github.com/ravendb/ravendb/pull/12243

File: test/RachisTests/DatabaseCluster/AtomicClusterReadWriteTests.cs [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+            using var store = GetDocumentStore(new Options {Server = leader, ReplicationFactor = nodes.Count});
+
+            var entity = new TestObj();
+            using (var session = store.OpenAsyncSession(new SessionOptions { TransactionMode = TransactionMode.ClusterWide }))
+            {
+                await session.StoreAsync(entity);
+                await session.SaveChangesAsync();
+            }
+
+            var result = await store.Operations.SendAsync(new GetCompareExchangeValuesOperation<User>(""));
```
Comment (ayende) [inline test/RachisTests/DatabaseCluster/AtomicClusterReadWriteTests.cs]: ??

File: test/RachisTests/DatabaseCluster/AtomicClusterReadWriteTests.cs [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+            using var store = GetDocumentStore(new Options {Server = leader, ReplicationFactor = nodes.Count});
+
+            var entity = new TestObj();
+            using (var session = store.OpenAsyncSession(new SessionOptions { TransactionMode = TransactionMode.ClusterWide, DisableAtomicDocumentWritesInClusterWideTransaction = true}))
+            {
+                await session.StoreAsync(entity);
+                await session.SaveChangesAsync();
+            }
+
+            var result = await store.Operations.SendAsync(new GetCompareExchangeValuesOperation<User>(""));
```
Comment (ayende) [inline test/RachisTests/DatabaseCluster/AtomicClusterReadWriteTests.cs]: ??

File: test/RachisTests/DatabaseCluster/AtomicClusterReadWriteTests.cs [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+
+            var entity = new TestObj();
+            using (var source = GetDocumentStore())
+            {
+                using var session = source.OpenAsyncSession(new SessionOptions { TransactionMode = TransactionMode.ClusterWide });
+                await session.StoreAsync(entity);
+                await session.SaveChangesAsync();
+
+                var operation = await source.Smuggler.ExportAsync(new DatabaseSmugglerExportOptions(), documentStore.Smuggler);
+                await operation.WaitForCompletionAsync(TimeSpan.FromMinutes(1));
```
Comment (ayende) [inline test/RachisTests/DatabaseCluster/AtomicClusterReadWriteTests.cs]: What is this for?

File: test/RachisTests/DatabaseCluster/AtomicClusterReadWriteTests.cs [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+            {
+                await session.StoreAsync(entity);
+                await session.SaveChangesAsync();
+            }
+
+            await LoadAndUpdateWhileDeleted(nodes, documentStore.Database, entity.Id);
+        }
+        
+        [Fact]
+        public async Task ClusterWideTransaction_WhenImportThenLoadAndUpdateWhileDeleted_ShouldSucceedOnlyInTheFirst()
```
Comment (ayende) [inline test/RachisTests/DatabaseCluster/AtomicClusterReadWriteTests.cs]: Effectively the same as `ClusterWideTransaction_WhenLoadAndUpdateWhileDeleted_ShouldSucceedOnlyInTheFirst`, no?

**Review by ayende — CHANGES_REQUESTED** [review]

**Comment by haludi** [comment]
> Isn't this supposed to be on top of: [ravendb#12243](https://github.com/ravendb/ravendb/pull/12243)

Yes, I tried to push to your branch with no success (as I understand this is the way to do that).
So I created a pull request to your branch.
How should I do that?

File: test/RachisTests/DatabaseCluster/AtomicClusterReadWriteTests.cs (hunk 1 above)
Comment (haludi) [reply test/RachisTests/DatabaseCluster/AtomicClusterReadWriteTests.cs]: Just checking a `CompareExchange` was created.
What does bother you here?
The empty string?

File: test/RachisTests/DatabaseCluster/AtomicClusterReadWriteTests.cs (hunk 2 above)
Comment (haludi) [reply test/RachisTests/DatabaseCluster/AtomicClusterReadWriteTests.cs]: https://github.com/ayende/ravendb/pull/4689/files/68b623274f57db5e63e9ec612115cea9c5faa169#r642026512

File: test/RachisTests/DatabaseCluster/AtomicClusterReadWriteTests.cs (hunk 3 above)
Comment (haludi) [reply test/RachisTests/DatabaseCluster/AtomicClusterReadWriteTests.cs]: I tried to think about edge cases.
An external source for a cluster-wide transaction document seemed to me as one?

File: test/RachisTests/DatabaseCluster/AtomicClusterReadWriteTests.cs (hunk 4 above)
Comment (haludi) [reply test/RachisTests/DatabaseCluster/AtomicClusterReadWriteTests.cs]: Maybe it is not necessary.
Just checking the same scenario after export import.
Should I remove it?

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 5]
```diff
@@ -543,145 +545,131 @@ private async ValueTask FinishBatchOfDocumentsAsync()
...
+                    using var doc = _database.DocumentsStorage.Get(_operationContext.ctx, docId);
+                    if (doc != null)
+                    {
+                        _commandParsedCommands.Push(new BatchRequestParser.CommandData
+                        {
+                            Id = docId,
+                            Document = doc.Data,
+                            Type = CommandType.PUT
+                        });
+                        doc.Data = null;
```
Comment (haludi) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: I did it to prevent dispose the data but it seems a little bit wired to me...

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 6]
```diff
@@ -543,145 +545,131 @@ private async ValueTask FinishBatchOfDocumentsAsync()
...
-                        await _database.TxMerger.Enqueue(updateClusterTransactionIndex);
-                        foreach (var docId in updateClusterTransactionIndex.CompareExchangeToRemove)
-                        {
-                            _compareExchangeRemoveCommands.Add(new RemoveCompareExchangeCommand(_database.Name,
-                                ClusterTransactionCommand.GetAtomicGuardKey(docId),0, context, RaftIdGenerator.DontCareId, fromBackup: true));
-                        }    
+                        _commandParsedCommands[i].Document.Dispose();
                     }
+                    _commandParsedCommands.Clear();
+                    await _database.ServerStore.Cluster.WaitForIndexNotification(ctcIndex);
```
Comment (haludi) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:668]: I wonder if we can `await` for this at the end of the import.
Also, maybe is worth `await` for changes to be applied on the current node (`_database.ClusterTransactionWaiter.WaitForResults(taskId, default);`).

**Comment by haludi** [comment]
I am also working on additional tests

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 7]
```diff
@@ -543,145 +545,131 @@ private async ValueTask FinishBatchOfDocumentsAsync()
...
             public async ValueTask WriteKeyValueAsync(string key, BlittableJsonReaderObject value)
             {
-                const int batchSize = 1024;
+                if(_operationContext.readTrx.Disposed)
+                    _operationContext.readTrx = _operationContext.ctx.OpenReadTransaction();
+
                 _compareExchangeAddOrUpdateCommands.Add(new AddOrUpdateCompareExchangeCommand(_database.Name, key, value, 0, _context, RaftIdGenerator.DontCareId, fromBackup: true));
+                if (ClusterTransactionCommand.IsAtomicGuardKey(key, out var docId))
+                {
+                    using var doc = _database.DocumentsStorage.Get(_operationContext.ctx, docId);
```
Comment (karmeli87) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: Is it okay to dispose the document here? maybe let's dispose it after sending the command?

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 8]
```diff
@@ -543,145 +545,131 @@ private async ValueTask FinishBatchOfDocumentsAsync()
...
             }
 
             public async ValueTask WriteKeyValueAsync(string key, BlittableJsonReaderObject value)
             {
-                const int batchSize = 1024;
+                if(_operationContext.readTrx.Disposed)
+                    _operationContext.readTrx = _operationContext.ctx.OpenReadTransaction();
+
                 _compareExchangeAddOrUpdateCommands.Add(new AddOrUpdateCompareExchangeCommand(_database.Name, key, value, 0, _context, RaftIdGenerator.DontCareId, fromBackup: true));
+                if (ClusterTransactionCommand.IsAtomicGuardKey(key, out var docId))
```
Comment (karmeli87) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:593]: Note that we don't want to add the compare exchange here if it is an atomic guard.

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 9]
```diff
@@ -543,145 +545,131 @@ private async ValueTask FinishBatchOfDocumentsAsync()
...
+                            Id = docId,
+                            Document = doc.Data,
+                            Type = CommandType.PUT
+                        });
+                        doc.Data = null;
+                    }
+                }
 
-                if (_compareExchangeAddOrUpdateCommands.Count < batchSize)
+                if (_compareExchangeAddOrUpdateCommands.Count < BatchSize)
```
Comment (karmeli87) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: Probably need to account for the `_commandParsedCommands` as well.

Comment (karmeli87) [reply src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: which is a wired name, ,maybe `_clusterTransactionCommands`?

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs (hunk 6 above)
Comment (karmeli87) [reply src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:668]: I think we should do `_database.ClusterTransactionWaiter.WaitForResults(taskId, import-token)`

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 10]
```diff
@@ -543,145 +545,131 @@ private async ValueTask FinishBatchOfDocumentsAsync()
...
-                            flags: document.Flags);
-                    }
-
-                    return count;
-                }
+                if (_compareExchangeAddOrUpdateCommands.Count == 0)
+                    return;
 
-                public override TransactionOperationsMerger.IReplayableCommandDto<TransactionOperationsMerger.MergedTransactionCommand> ToDto(JsonOperationContext context)
+                var (aoucecIndex, _) = await _database.ServerStore.SendToLeaderAsync(new AddOrUpdateCompareExchangeBatchCommand(_compareExchangeAddOrUpdateCommands, context, RaftIdGenerator.DontCareId));
```
Comment (karmeli87) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:645]: `aoucec` ?

**Review by karmeli87 — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs (hunk 7 above)
Comment (haludi) [reply src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: Why should we hold unused data?

Comment (karmeli87) [reply src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: why unused? you need it for the cluster-tx, no?

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs (hunk 10 above)
Comment (haludi) [reply src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:645]: **A**dd**O**r**U**pdate**C**ompare**E**xchange**C**ommand
a___o__u______c_________e________c

Comment (karmeli87) [reply src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:645]: let's call it simply `index` or do `var result = ... `

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 11]
```diff
@@ -585,26 +590,26 @@ public DisposableReturnedArray(int length)
...
             {
-                if(_operationContext.readTrx.Disposed)
-                    _operationContext.readTrx = _operationContext.ctx.OpenReadTransaction();
-
-                _compareExchangeAddOrUpdateCommands.Add(new AddOrUpdateCompareExchangeCommand(_database.Name, key, value, 0, _context, RaftIdGenerator.DontCareId, fromBackup: true));
                 if (ClusterTransactionCommand.IsAtomicGuardKey(key, out var docId))
                 {
+                    if (_operationContext.readTrx == null || _operationContext.readTrx.Disposed)
+                        _operationContext.readTrx = _operationContext.ctx.OpenReadTransaction();
+
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:597]: Need to drop the using below

**Review by ayende — APPROVED** [review]

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs (hunk 11 above)
Comment (haludi) [reply src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:597]: What do you mean?
In line 652?

Comment (haludi) [reply src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:597]: What do you mean?
We

Comment (ayende) [reply src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:597]: No, line 598 - but you already did that in a later commir
